### PR TITLE
Add new sampling algorithm largest triangle to sample UDF

### PR DESCRIPTION
**Pull request by lizhizhou**
## Description
Add the lagest triangle sampling algorithm to iotdb UDF.
The algorithm is verified in porduction IoT system for a while and the result show it's a very useful algorithm for visualization 

### Content3 ...

<hr>

This PR has:
- [ ] been self-reviewed.
    - [ ] concurrent read
    - [ ] concurrent write
    - [ ] concurrent read and write 
- [ ] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. 
- [ ] added or updated version, __license__, or notice information
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious 
  for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold 
  for code coverage.
- [ ] added integration tests.
- [ ] been tested in a test IoTDB cluster.

<hr>

### CODE REVIEW

File: library-udf/src/main/java/org/apache/iotdb/library/dprofile/UDTFSample.java [hunk 1]
```diff
@@ -105,11 +123,55 @@ public void transform(Row row, PointCollector collector) throws Exception {
...
     if (this.k < n) {
-      for (long i = 0; i < this.k; i++) {
-        long j = Math.floorDiv(i * (long) n, (long) k); // avoid intermediate result overflows
-        Row row = rowWindow.getRow((int) j);
-        Util.putValue(collector, dataType, row.getTime(), Util.getValueAsObject(row));
+      if (this.method == Method.TRIANGLE)
+      {
+        System.out.println("use triangle");
+        List<Point> input = new LinkedList<>();
+        for (long i = 0; i < n; i++) {
```
Comment (JackieTien97) [inline library-udf/src/main/java/org/apache/iotdb/library/dprofile/UDTFSample.java:132]:
```suggestion
        for (int i = 0; i < n; i++) {
```

File: library-udf/src/main/java/org/apache/iotdb/library/dprofile/UDTFSample.java [hunk 2]
```diff
@@ -105,11 +123,55 @@ public void transform(Row row, PointCollector collector) throws Exception {
...
+          Row row = rowWindow.getRow((int) i);
+          BigDecimal time = new BigDecimal(row.getTime());
+          BigDecimal data = new BigDecimal(Util.getValueAsDouble(row));
```
Comment (JackieTien97) [inline library-udf/src/main/java/org/apache/iotdb/library/dprofile/UDTFSample.java:135]:
```suggestion
          BigDecimal data = BigDecimal.valueOf(Util.getValueAsDouble(row));
```

Comment (JackieTien97) [reply library-udf/src/main/java/org/apache/iotdb/library/dprofile/UDTFSample.java:135]: The previous way is an unpredictable 'new BigDecimal()' call.

File: library-udf/src/main/java/org/apache/iotdb/library/dprofile/UDTFSample.java [hunk 3]
```diff
@@ -105,11 +123,55 @@ public void transform(Row row, PointCollector collector) throws Exception {
...
+                break;
+              case DOUBLE:
+                Util.putValue(collector, dataType, p.getX().longValue(), p.getY().doubleValue());
+                break;
+              default:
+                throw new NoNumberException();
+            }
+          }
+        } else { //For corner case of k == 1 and k == 2
+          Row row = rowWindow.getRow((int) 0); // Put first element
```
Comment (JackieTien97) [inline library-udf/src/main/java/org/apache/iotdb/library/dprofile/UDTFSample.java:161]:
```suggestion
          Row row = rowWindow.getRow(0); // Put first element
```

File: library-udf/src/main/java/org/apache/iotdb/library/dprofile/UDTFSample.java [hunk 4]
```diff
@@ -105,11 +123,55 @@ public void transform(Row row, PointCollector collector) throws Exception {
...
+          Util.putValue(collector, dataType, row.getTime(), Util.getValueAsObject(row));
+          if (k == 2) {
+            row = rowWindow.getRow((int) n-1); // Put last element
```
Comment (JackieTien97) [inline library-udf/src/main/java/org/apache/iotdb/library/dprofile/UDTFSample.java:164]:
```suggestion
            row = rowWindow.getRow(n - 1); // Put last element
```

File: library-udf/src/main/java/org/apache/iotdb/library/dprofile/UDTFSample.java [hunk 5]
```diff
@@ -105,11 +123,55 @@ public void transform(Row row, PointCollector collector) throws Exception {
...
+          Row row = rowWindow.getRow((int) i);
+          BigDecimal time = new BigDecimal(row.getTime());
+          BigDecimal data = new BigDecimal(Util.getValueAsDouble(row));
+          input.add(new Point(time, data));
+        }
+        if (k > 2) {
+          // The first and last element will always be sampled so the buckets is k - 2
+          List<Point> output = LTThreeBuckets.sorted(input, k - 2);
+          for (Point p : output) {
+            System.out.println("point: " + p);
```
Comment (JackieTien97) [inline library-udf/src/main/java/org/apache/iotdb/library/dprofile/UDTFSample.java:142]: use logger.debug instead.

**Review by JackieTien97 — CHANGES_REQUESTED** [review]
Remember add docs to describe this fantastic UDF in `docs/zh/UserGuide/UDF-Library/Data-Profiling.md` and `docs/UserGuide/UDF-Library/Data-Profiling.md`.

**Comment by lizhizhou** [comment]
Resolve all the comments and add documents

File: library-udf/pom.xml [hunk 6]
```diff
@@ -27,6 +27,23 @@
...
+            </releases>
+            <snapshots>
+                <enabled>true</enabled>
+            </snapshots>
+        </repository>
+        <repository>
+            <id>jitpack.io</id>
+            <url>https://jitpack.io</url>
+        </repository>
+    </repositories>
```
[the remark this answers is not in the document]
Comment (jixuan1989) [reply library-udf/pom.xml:46]: well, we must add comments to show which dependency uses this repository.
If it is really necessary, it is ok.
BUT,  apache nexus snapshot repo should not be included.

**Review by jixuan1989 — CHANGES_REQUESTED** [review]
pls remove apache.snapshots repository

**Comment by lizhizhou** [comment]
All comment resolved

**Pull request merged**